Assets/Mirror/Runtime/NetworkDiagnostics.cs

```diff
@@ -0,0 +1,78 @@
+using System;
+using System.Diagnostics;
+using UnityEngine.Events;
+
+namespace Mirror
+{
+    /// <summary>
+    /// Provides profiling information from mirror
+    /// A profiler can subscribe to these events and
+    /// present the data in a friendly way to the user
+    /// </summary>
+    public static class NetworkDiagnostics
+    {
+        /// <summary>
+        /// Describes an outgoing message
+        /// </summary>
+        public readonly struct MessageInfo
+        {
+            /// <summary>
+            /// The message being sent
+            /// </summary>
+            public readonly IMessageBase message;
+            /// <summary>
+            /// channel through which the message was sent
+            /// </summary>
+            public readonly int channel;
+            /// <summary>
+            /// how big was the message (does not include transport headers)
+            /// </summary>
+            public readonly int bytes;
+            /// <summary>
+            /// How many connections was the message sent to
+            /// If an object has a lot of observers this count could be high
+            /// </summary>
+            public readonly int count;
+
+            internal MessageInfo(IMessageBase message, int channel, int bytes, int count)
+            {
+                this.message = message;
+                this.channel = channel;
+                this.bytes = bytes;
+                this.count = count;
+            }
+        }
+
+        #region Out messages
+        /// <summary>
+        /// Event that gets raised when Mirror sends a message
+        /// Subscribe to this if you want to profile the network
+        /// </summary>
+        public static event Action<MessageInfo> OutMessageEvent;
+
+        [Conditional("ENABLE_PROFILER")]
+        internal static void OnSend<T>(T message, int channel, int bytes, int count) where T : IMessageBase
+        {
+            MessageInfo outMessage = new MessageInfo(message, channel, bytes, count);
+            OutMessageEvent?.Invoke(outMessage);
+        }
+        #endregion
+
+        #region In messages
+
+        /// <summary>
+        /// Event that gets raised when Mirror receives a message
+        /// Subscribe to this if you want to profile the network
+        /// </summary>
+        public static event Action<MessageInfo> InMessageEvent;
+        
+        [Conditional("ENABLE_PROFILER")]
```

Comment:
```suggestion
        [Conditional("ENABLE_DIAGNOSTICS")]
```

Reply:
Unity does not turn on ENABLE_DIAGNOSTICS out of the box.   But i will will turn on ENABLE_PROFILER if you enable profiling,   this is not my define,  it is Unity's